### PR TITLE
[TG-4331] Support multiple types for input parameters for entry point function

### PR DESCRIPTION
**Pull request by mgudemann**
The pointer selector can return a set of types to use for the parameters of the entry point function. This will create a non-deterministic choice between the different types in `__CPROVER__start`.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f18351b48f784bc30c4beca984fb750b0e308ce4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80864928

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 1]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
-    main_arguments[param_number] = object_factory(
-      p.type(),
-      base_name,
-      init_code,
-      symbol_table,
-      parameters,
-      allocation_typet::LOCAL,
-      function.location,
-      pointer_type_selector);
+    // argument, if the argument has different possible object types (from casts
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: The comment `from casts in the function body` sounds more like a possible heuristics than a necessity.

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 2]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
+      //     break
+      //   ...
+      // }
+      // method(..., param, ...)
+      //
+
+      const symbolt result_symbol = get_fresh_aux_symbol(
+        p.type(),
+        "nondet_parameter_" + std::to_string(param_number),
+        "nondet_parameter_" + std::to_string(param_number),
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: This takes the current function identifier.

File: jbmc/src/java_bytecode/nondet.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  symbol_table_baset &symbol_table,
+  code_blockt &instructions)
+{
+  PRECONDITION(min_value < max_value);
+
+  // Declare a symbol for the non deterministic integer.
+  const symbol_exprt &nondet_symbol = get_fresh_aux_symbol(
+                                        int_type,
+                                        name_prefix + "::nondet_int",
+                                        "nondet_int",
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/nondet.cpp]: must take the function id

File: jbmc/src/java_bytecode/nondet.cpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    id2string(name_prefix),
+    int_type,
+    source_location,
+    symbol_table,
+    result_block);
+
+  result_switch.value() = switch_value;
+
+  code_blockt switch_block;
+  int64_t case_number = 0;
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/nondet.cpp:115]: `size_t`

File: jbmc/src/java_bytecode/nondet.cpp [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+/// \param name_prefix: Prefix for the fresh symbol name generated.
+/// \param int_type: The type of the int used to non-deterministically choose
+///   one of the switch cases.
+/// \param source_location: The location to mark the generated int with.
+/// \param symbol_table: The global symbol table.
+/// \param instructions [out]: Output instructions are written to
+///   'instructions'. These declare, nondet-initialise and range-constrain (with
+///   assume statements) a fresh integer.
+/// \return Returns a symbol expression for the resulting integer.
+symbol_exprt generate_nondet_int(
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/nondet.cpp:35]: This could possibly be generalised and moved out of JBMC, but I suggest to do this in a separate PR.

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 6]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
+      // switch(...)
+      // {
+      //   case obj1:
+      //     tmp_expr = object_factory(...)
+      //     param = tmp_expr
+      //     break
+      //   ...
+      // }
+      // method(..., param, ...)
+      //
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_entry_point.cpp:363]: I don't like long comments in the middle of a function, could you extract the relevant part of the code in a function and make that comment its documentation?

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 7]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
+        ID_java,
+        symbol_table);
+      main_arguments[param_number] = result_symbol.symbol_expr();
+
+      std::vector<codet> cases;
+      size_t alternative = 0;
+      for(const auto &type : alternative_object_types)
+      {
+        code_blockt init_code_for_type;
+        exprt init_expr_for_parameter = object_factory(
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: const

**Review by peterschrammel — CHANGES_REQUESTED** [review]

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 8]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
+          java_reference_type(type),
+          id2string(base_name) + "_alt_" + std::to_string(alternative),
+          init_code_for_type,
+          symbol_table,
+          parameters,
+          allocation_typet::DYNAMIC,
+          function.location,
+          pointer_type_selector);
+        alternative++;
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: could we make the for loop on alternative to avoid this increment in the middle

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 9]
```diff
@@ -317,17 +319,89 @@ exprt::operandst java_build_arguments(
...
+        init_code_for_type.add(
+          code_assignt(
+            result_symbol.symbol_expr(),
+            typecast_exprt(init_expr_for_parameter, p.type())));
+        cases.push_back(init_code_for_type);
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: alternatively we could use transform instead of this for loop

File: jbmc/src/java_bytecode/nondet.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+symbol_exprt generate_nondet_int(
+  int64_t min_value,
+  int64_t max_value,
+  const std::string &name_prefix,
+  const typet &int_type,
+  const source_locationt &source_location,
+  symbol_table_baset &symbol_table,
+  code_blockt &instructions);
+
+typedef std::vector<codet> alternate_casest;
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/nondet.h:25]: this typedef is not necessary (the semantic of the vector should be given in the variable name rather than the type)

File: jbmc/src/java_bytecode/nondet.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+/// Pick nondeterministically between imperative actions 'switch_cases'.
+/// \param name_prefix: Name prefix for fresh symbols
+/// \param switch_cases: List of codet objects to execute in each switch case.
+/// \param int_type: The type of the int used to non-deterministically choose
+///   one of the switch cases.
+/// \param source_location: The location to mark the generated int with.
+/// \param symbol_table: The global symbol table.
+/// \return Returns a nondet-switch choosing between switch_cases. The resulting
+///   switch block has no default case.
+code_blockt generate_nondet_switch(
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/nondet.cpp:88]: `generate` in the name not necessary

File: jbmc/src/java_bytecode/select_pointer_type.cpp [hunk 12]
```diff
@@ -222,3 +222,12 @@ select_pointer_typet::get_recursively_instantiated_type(
...
   return inst_val;
 }
+
+optionalt<std::set<symbol_typet>>
+select_pointer_typet::get_parameter_alternative_types(
+  const irep_idt &function_name,
+  const irep_idt &parameter_name,
+  const namespacet &ns) const
+{
+  return {};
```
Comment (romainbrenguier) [inline jbmc/src/java_bytecode/select_pointer_type.cpp:232]: I don't seen any other implementation of this method, so I don't really see its purpose.

File: jbmc/src/java_bytecode/nondet.cpp [hunk 13]
```diff
@@ -0,0 +1,125 @@
+/// Author: Diffblue Limited. All Rights Reserved.
+
+#include "nondet.h"
+
+#include <java_bytecode/java_types.h>
```
Comment (tautschnig) [inline jbmc/src/java_bytecode/nondet.cpp]: I don't see where this would be used in this file, and, really, this should not live in the `java_bytecode` folder - it's way too big already.

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: jbmc/src/java_bytecode/nondet.cpp (hunk 5 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/nondet.cpp:35]: the whole file is taken from TG, should we do all suggested changes in a separate PR ?

File: jbmc/src/java_bytecode/select_pointer_type.cpp (hunk 12 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/select_pointer_type.cpp:232]: it's currently only implemented in smart pointer selection, this will later be ported to JBMC

File: jbmc/src/java_bytecode/nondet.cpp (hunk 13 above)
Comment (mgudemann) [reply jbmc/src/java_bytecode/nondet.cpp]: @tautschnig I see your point, but this is moved from TG to avoid duplicating code unneccesarily. Making this more general and non-Java specific is suggested by @peterschrammel too, but it should maybe  be a separate PR.

Comment (tautschnig) [reply jbmc/src/java_bytecode/nondet.cpp]: It's not particularly clear to me what is actually Java-specific in this file? I thought it could be moved to a different folder with the single modification of dropping this (unnecessary) include?

**Comment by mgudemann** [comment]
@tautschnig @peterschrammel please have another look at this. The `nondet` file is copied from TG, I agree it should be generalized and maybe moved, but currently it only supports Java. The functionality is also currently not implemented using the simple pointer selection mechanism, but TG uses it, putting this into the pointer selection was what I agreed with @peterschrammel

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 14]
```diff
@@ -317,17 +319,94 @@ exprt::operandst java_build_arguments(
...
+      //     param = tmp_expr
+      //     break
+      //   ...
+      // }
+      // method(..., param, ...)
+      //
+
+      const symbolt result_symbol = get_fresh_aux_symbol(
+        p.type(),
+        "nondet_parameter_" + id2string(function.name) + "_" +
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_entry_point.cpp]:
```
"nondet_parameter_" + std::to_string(param_number),
id2string(function.name),
```

File: jbmc/src/java_bytecode/java_entry_point.cpp [hunk 15]
```diff
@@ -317,17 +319,94 @@ exprt::operandst java_build_arguments(
...
+
+      std::transform(
+        alternatives.begin(),
+        alternatives.end(),
+        cases.begin(),
+        initialize_parameter);
+
+      init_code.add(
+        generate_nondet_switch(
+          id2string(function.base_name) + "_" + std::to_string(param_number),
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/java_entry_point.cpp]: no, see above.

**Review by peterschrammel — CHANGES_REQUESTED** [review]

**Review by tautschnig — APPROVED** [review]

**Comment by tautschnig** [comment]
Approving with the assumption that there will be a more-or-less immediate follow-up PR that moves nondet.{h,cpp} into a proper place.

**Review by peterschrammel — APPROVED** [review]
